File: backend/src/services/todo_service.py
from sqlmodel import Session, select, update
from typing import List, Optional
from ..models.todo import Todo, TodoCreate, TodoUpdate
from ..models.user import User
from datetime import datetime
import uuid


def create_todo(session: Session, todo: TodoCreate, user_id: uuid.UUID) -> Todo:
    """Create a new todo for a user."""
    db_todo = Todo.model_validate(todo)
    db_todo.user_id = user_id
    session.add(db_todo)
    session.commit()
    session.refresh(db_todo)
    return db_todo


def get_todos_by_user(
    session: Session,
    user_id: uuid.UUID,
    completed: Optional[bool] = None,
    offset: int = 0,
    limit: int = 100
) -> List[Todo]:
    """Get all todos for a specific user, with optional filtering."""
    query = select(Todo).where(Todo.user_id == user_id)

    if completed is not None:
        query = query.where(Todo.is_completed == completed)

    query = query.offset(offset).limit(limit)

    return session.exec(query).all()


def get_todo_by_id_and_user(session: Session, todo_id: uuid.UUID, user_id: uuid.UUID) -> Optional[Todo]:
    """Get a specific todo by ID for a specific user."""
    query = select(Todo).where(Todo.id == todo_id, Todo.user_id == user_id)
    return session.exec(query).first()


def update_todo_by_id_and_user(
    session: Session,
    todo_id: uuid.UUID,
    user_id: uuid.UUID,
    todo_update: TodoUpdate
) -> Optional[Todo]:
    """Update a specific todo by ID for a specific user."""
    db_todo = get_todo_by_id_and_user(session, todo_id, user_id)
    if not db_todo:
        return None

    # Prepare update data
    update_data = todo_update.model_dump(exclude_unset=True)

    # Update the todo
    for field, value in update_data.items():
        setattr(db_todo, field, value)

    db_todo.updated_at = datetime.utcnow()

    session.add(db_todo)
    session.commit()
    session.refresh(db_todo)

    return db_todo


def delete_todo_by_id_and_user(session: Session, todo_id: uuid.UUID, user_id: uuid.UUID) -> bool:
    """Delete a specific todo by ID for a specific user."""
    db_todo = get_todo_by_id_and_user(session, todo_id, user_id)
    if not db_todo:
        return False

    session.delete(db_todo)
    session.commit()
    return True


def toggle_todo_completion(session: Session, todo_id: uuid.UUID, user_id: uuid.UUID) -> Optional[Todo]:
    """Toggle the completion status of a specific todo."""
    db_todo = get_todo_by_id_and_user(session, todo_id, user_id)
    if not db_todo:
        return None

    db_todo.is_completed = not db_todo.is_completed
    db_todo.updated_at = datetime.utcnow()

    session.add(db_todo)
    session.commit()
    session.refresh(db_todo)

    return db_todo


def validate_user_owns_resource(session: Session, user_id: uuid.UUID, resource_user_id: uuid.UUID) -> bool:
    """Validate that the authenticated user owns the requested resource."""
    return str(user_id) == str(resource_user_id)

File: lib/todoService.ts
// frontend/lib/todoService.ts

import apiClient, { todoAPI } from './api';
import { Todo, TodoCreate, TodoUpdate } from '@/types';

// Service functions for todo operations
export const todoService = {
  // Get all todos for the authenticated user
  getTodos: async (): Promise<Todo[]> => {
    try {
      const response = await todoAPI.getTodos();
      return response.data;
    } catch (error) {
      console.error('Error fetching todos:', error);
      throw error;
    }
  },

  // Create a new todo
  createTodo: async (todoData: TodoCreate): Promise<Todo> => {
    try {
      const response = await todoAPI.createTodo(todoData);
      return response.data;
    } catch (error) {
      console.error('Error creating todo:', error);
      throw error;
    }
  },

  // Get a specific todo
  getTodo: async (todoId: string): Promise<Todo> => {
    try {
      const response = await todoAPI.getTodo(todoId);
      return response.data;
    } catch (error) {
      console.error(`Error fetching todo with id ${todoId}:`, error);
      throw error;
    }
  },

  // Update a specific todo
  updateTodo: async (todoId: string, todoData: TodoUpdate): Promise<Todo> => {
    try {
      const response = await todoAPI.updateTodo(todoId, todoData);
      return response.data;
    } catch (error) {
      console.error(`Error updating todo with id ${todoId}:`, error);
      throw error;
    }
  },

  // Toggle completion status of a todo
  toggleTodoComplete: async (todoId: string): Promise<Todo> => {
    try {
      const response = await todoAPI.toggleTodoComplete(todoId);
      return response.data;
    } catch (error) {
      console.error(`Error toggling todo completion with id ${todoId}:`, error);
      throw error;
    }
  },

  // Delete a specific todo
  deleteTodo: async (todoId: string): Promise<void> => {
    try {
      await todoAPI.deleteTodo(todoId);
    } catch (error) {
      console.error(`Error deleting todo with id ${todoId}:`, error);
      throw error;
    }
  },
};